### PR TITLE
CFE-3415: Fixed some more sign-compare warnings

### PR DESCRIPTION
**Pull request by larsewi**

### CODE REVIEW

File: cf-agent/files_editline.c [hunk 1]
```diff
@@ -1245,7 +1245,7 @@ static int ReplacePatterns(EvalContext *ctx, Item *file_start, Item *file_end, c
 
         while (BlockTextMatch(ctx, pp->promiser, line_buff, &start_off, &end_off))
         {
-            if (match_len == strlen(line_buff))
+            if (match_len == (int) strlen(line_buff))
```
Comment (vpodzime) [inline cf-agent/files_editline.c:1248]: Could `match_len` be `size_t`?

File: cf-agent/verify_users_pam.c [hunk 2]
```diff
@@ -1136,79 +1145,101 @@ static void TransformGidsToGroups(StringSet **list)
...
     {
-        CFUSR_SETBIT (*changemap, i_comment);
+        CFUSR_SETBIT(*changemap, i_comment);
     }
-    if (u->uid != NULL && (atoi (u->uid) != passwd_info->pw_uid))
+
+    errno = 0;
+    if (u->uid != NULL && 
+        (uid_t) strtoul(u->uid, (char **) NULL, 10) != passwd_info->pw_uid &&
+        errno != 0)
```
Comment (vpodzime) [inline cf-agent/verify_users_pam.c:1166]: The `errno` comparison here is confusing at least and also wrong, I think.

File: libpromises/dbm_lmdb.c [hunk 3]
```diff
@@ -248,12 +248,17 @@ static bool RepairedAfterOpen(const char *lmdb_file, int fd_tstamp)
...
+    {
+        Log(LOG_LEVEL_ERR, "Failed to read %s: %s", lmdb_file, GetErrorStr());
+    }
+    else if (n_read == 0)
     {
         /* EOF (empty file) => never repaired */
         Log(LOG_LEVEL_VERBOSE, "DB '%s' never repaired before", lmdb_file);
     }
-    else if (n_read < sizeof(time_t))
+    else if ((unsigned long) n_read < sizeof(time_t))
```
Comment (vpodzime) [inline libpromises/dbm_lmdb.c:261]: Why `unsigned long` here?

**Review by vpodzime — COMMENTED** [review]
I think we should try to not rely on `errno` after calling `strtoul()`. Let implement a `StringToUlong()` function that would use `strtoul()` and in case of `strtoul()` returns 0 it would check if the string was `"0"` (or any number of zeroes).

File: libpromises/dbm_lmdb.c (hunk 3 above)
Comment (larsewi) [reply libpromises/dbm_lmdb.c:261]: My bad. Will fix!

File: cf-agent/verify_users_pam.c [hunk 4]
```diff
@@ -991,7 +991,16 @@ static bool GetGroupInfo (const char *user, const User *u, StringSet **groups_to
...
 {
     assert(entry != NULL);
-    return (atoi(key) == entry->gr_gid);
+
+    errno = 0;
+    gid_t gid = (gid_t) strtoul(key, (char **) NULL, 10);
+    if (errno != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to read group id from key: %s", GetErrorStr());
+        return false;
```
Comment (larsewi) [inline cf-agent/verify_users_pam.c:1000]: Should this return false on error or should it exit on error?

Comment (olehermanse) [reply cf-agent/verify_users_pam.c:1000]: When would this happen? Generally speaking we should not exit on errors. The exception is programming errors / asserts in debug builds, and critical errors (with a critical log message) which prevent CFEngine from functioning.

File: cf-agent/files_editline.c (hunk 1 above)
Comment (larsewi) [reply cf-agent/files_editline.c:1248]: Yes you're right ! 👍

**Comment by larsewi** [comment]
@cf-bottom jenkins, please!

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6062/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6062/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6062/

File: cf-agent/files_editline.c [hunk 5]
```diff
@@ -1656,11 +1657,26 @@ static bool MatchPolicy(EvalContext *ctx, const char *camel, const char *haystac
                     char  *work      = xcalloc(1, work_size);
                     strcpy(work, sp);
 
-                    if (snprintf(final, final_size, "\\s*%s", work) >= final_size - 1)
+                    int written = snprintf(final, final_size, "\\s*%s", 
+                                           work);
+                    if (written < 0)
+                    {
+                        Log(LOG_LEVEL_ERR, "snprintf failed: %s",
```
Comment (vpodzime) [inline cf-agent/files_editline.c:1664]: Imagine debugging a failure like this. All you'll get is `snprintf failed: SOME REASON` and there will be no clue where and what the `snprintf()` was even trying to format. We need to add the information like that, preferably with some wording understandable to the users if they are likely to see these errors.

File: libpromises/evalfunction.c [hunk 6]
```diff
@@ -1953,37 +1953,37 @@ static FnCallResult FnCallTextXform(ARG_UNUSED EvalContext *ctx, ARG_UNUSED cons
     char *buf = xcalloc(bufsiz, sizeof(char));
     memcpy(buf, string, len + 1);
 
-    if (!strcmp(fp->name, "string_downcase"))
+    if (strcmp(fp->name, "string_downcase") == 0)
```
Comment (vpodzime) [inline libpromises/evalfunction.c:1956]: `StringEqual()`

Comment (vpodzime) [reply libpromises/evalfunction.c:1956]: Applies to all the `strcmp() == 0` comparisons below.

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise. Thanks for so many nice changes!!! :clap:

**Review by vpodzime — APPROVED** [review]
Looks good to me. Thanks!

**Pull request merged**